### PR TITLE
Fix calling FlushAsync with cancelled token

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/dotnet/corefx/issues/28492

We were marking awaitable as completed but never calling the callback because `AttachToken` threw.

Reorder calls to allow it to throw before completing awaitable.

### CODE REVIEW

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs [hunk 1]
```diff
@@ -307,6 +307,35 @@ public void GetResultThrowsIfFlushAsyncTokenFiredAfterCancelPending()
...
+            cancellationTokenSource.Cancel();
+
+            // AsTask is important here, it validates that we are calling completion callback
+            // and not only setting IsCompleted flag
+            var task = Pipe.Reader.ReadAsync().AsTask();
+            bool threwException = false;
+
+            try
+            {
+                await Pipe.Writer.FlushAsync(cancellationTokenSource.Token);
```
Comment (halter73) [inline src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: Why not Assert.ThrowsAsync?

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs [hunk 2]
```diff
@@ -307,6 +307,35 @@ public void GetResultThrowsIfFlushAsyncTokenFiredAfterCancelPending()
...
+            {
+                threwException = true;
+            }
+            finally
+            {
+                Pipe.Writer.Complete();
+            }
+
+            Assert.True(threwException);
+            Assert.True(task.IsCompleted);
```
Comment (halter73) [inline src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: I think IsCompleted is true for faulted/cancelled tasks, so this is guaranteed to be true.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 3]
```diff
@@ -298,10 +298,10 @@ internal ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
             {
                 CommitUnsynchronized();
 
-                _readerAwaitable.Complete(out completionData);
-
                 cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, s_signalWriterAwaitable, this);
 
+                _readerAwaitable.Complete(out completionData);
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Can you please add a comment stating why this ordering is necessary for correctness?

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs (hunk 2 above)
Comment (halter73) [reply src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: Nvm. I didn't read this test correctly. I now see this is the read task.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 4]
```diff
@@ -298,10 +298,10 @@ internal ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
             {
                 CommitUnsynchronized();
 
-                _readerAwaitable.Complete(out completionData);
-
                 cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, s_signalWriterAwaitable, this);
```
Comment (halter73) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Why are we ever disposing this registration below before the returned ValueTask completes?

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Because that's the old registration we replaced.

**Comment by pakrym** [comment]
@dotnet-bot test OSX x64 Debug Build

**Pull request merged**